react-todo-props/src/hooks/useTodoService.js

```diff
@@ -0,0 +1,49 @@
+import {v4 as uuid} from "uuid";
+import {useEffect, useState} from "react";
+
+export class TodoStorage {
+    STORAGE_KEY = 'todoList';
+
+    getList() {
+        const result = JSON.parse(localStorage.getItem(this.STORAGE_KEY));
+        return result ? result : [];
+    }
+
+    save(todoList) {
+        localStorage.setItem(this.STORAGE_KEY, JSON.stringify(todoList));
+    }
+}
+
+export const useTodoService = (todoStorage) => {
+    const [todoList, setTodoList] = useState(getList());
+
+    useEffect(() => {
+        save([...todoList]);
+    }, [todoList]);
+
+    function saveTodo(todo) {
+        const addTodo = {id: uuid(), checked: false, editState: false, ...todo};
+        setTodoList([...todoList, addTodo]);
+    }
+
+    function removeTodo(todo) {
+        todoList.splice(todoList.indexOf(todo), 1);
+        setTodoList([...todoList]);
+    }
+
+    function updateTodo(old, newTodo) {
+        const result = [...todoList];
+        result.splice(result.indexOf(old), 1, {...old, ...newTodo});
+        setTodoList(result);
+    }
+
+    function getList() {
+        return todoStorage.getList();
+    }
+
+    function save(todoList) {
+        todoStorage.save(todoList);
+    }
```

Comment:
이곳에 함수표현식이 아닌 함수선언식으로 사용하신 이유가 있으실까요 ?

Reply:
```js
const [todoList, setTodoList] = useState(getList());
```
이 부분을 함수 선언 최상단에 올리고 싶은데 표현식으로 선언하면 const로 선언한 저 부분을 아래로 내려야 해서 function으로 변경했어요.! 다른 방법이 있는지 궁금해요~!

Reply:
```suggestion

export const useTodoService = () => {
  const STORAGE_KEY = 'todoList';

  const getList = () => {
     const result = JSON.parse(localStorage.getItem(STORAGE.STORAGE_KEY));
     return result;
   };

  const save = (todoList) => {
    localStorage.setItem(STORAGE.STORAGE_KEY, JSON.stringify(todoList));
  };

  const [todoList, setTodoList] = useState(getList());

  useEffect(() => {
    save([...todoList]);
  }, [todoList]);

  const saveTodo = (todo) => {
    const addTodo = { id: uuid(), checked: false, editState: false, ...todo };
    setTodoList([...todoList, addTodo]);
  };

  const removeTodo = (todo) => {
    todoList.splice(todoList.indexOf(todo), 1);
    setTodoList([...todoList]);
  };

  const updateTodo = (old, newTodo) => {
    const result = [...todoList];
    result.splice(result.indexOf(old), 1, { ...old, ...newTodo });
    setTodoList(result);
  };

  return { todoList, saveTodo, removeTodo, updateTodo };
};

```

Reply:
TodoStorage를 커스텀훅에 class로 선언하신 이유가 있으실까요 ?
localStorage만 다루는 파일을 분리해서 재사용성에 대해서 고려해보시는것도 좋으실 것 같습니다.
서비스 로직의 위치도 고려하셔서 위치를 다시한번 고려해보시면 좋으실 것 같습니다.
커스텀 훅의 규칙에 대해서도 알아보시면 좋으실 것 같습니다!
고생하셨습니다.

Reply:
```suggestion

export const TodoService = () => {
  const STORAGE_KEY = 'todoList';
  const getList = () => {
    return JSON.parse(localStorage.getItem(STORAGE_KEY));
  };

  const save = (todoList) => {
    localStorage.setItem(STORAGE_KEY, JSON.stringify(todoList));
  };
  const [todoList, setTodoList] = useState(getList());

  useEffect(() => {
    save([...todoList]);
  }, [todoList]);

  const saveTodo = (todo) => {
    const addTodo = { id: uuid(), checked: false, editState: false, ...todo };
    setTodoList([...todoList, addTodo]);
  };

  const removeTodo = (todo) => {
    todoList.splice(todoList.indexOf(todo), 1);
    setTodoList([...todoList]);
  };

  const updateTodo = (old, newTodo) => {
    const result = [...todoList];
    result.splice(result.indexOf(old), 1, { ...old, ...newTodo });
    setTodoList(result);
  };

  return { todoList, saveTodo, removeTodo, updateTodo };
};

```

TodoStorage를 커스텀훅에 class로 선언하신 이유가 있으실까요 ?
localStorage만 다루는 파일을 분리해서 재사용성에 대해서 고려해보시는것도 좋으실 것 같습니다.
서비스 로직의 위치도 고려하셔서 위치를 다시한번 고려해보시면 좋으실 것 같습니다.
커스텀 훅의 규칙에 대해서도 알아보시면 좋으실 것 같습니다!
고생하셨습니다.